addons/l10n_it_edi/models/account_edi_format.py

```diff
@@ -72,9 +72,21 @@ def _l10n_it_edi_is_self_invoice(self, invoice):
             We recognize these cases based on the taxes that target the VJ tax grids, which imply
             the use of VAT External Reverse Charge.
         """
-        report_lines_xmlids = invoice.line_ids.tax_tag_ids.tax_report_line_ids.mapped(lambda x: x.get_external_id().get(x.id, ''))
-        return (invoice.is_purchase_document()
-                and any([x.startswith("l10n_it.tax_report_line_vj") for x in report_lines_xmlids]))
+        if not invoice.is_purchase_document():
+            return False
+
+        invoice_lines_tag_ids = invoice.line_ids.tax_tag_ids.ids
+        it_tax_report_id = self.env.ref('l10n_it.tax_report_vat').id
+        it_tax_report_expressions = self.env['account.report.expression'].search([('report_line_id.report_id', '=', it_tax_report_id)])
+        for expression in it_tax_report_expressions:
+            expression_xml_id = expression.get_external_id().get(expression.id, '')
+            if not expression_xml_id.startswith('l10n_it.tax_report_line_vj'):
+                continue
+
+            matching_tags_ids = expression._get_matching_tags().ids
+            if any(tag_id in invoice_lines_tag_ids for tag_id in matching_tags_ids):
+                return True
```

Comment:
report_lines = self.env['account.report.line'].search([('report_id.country_id.code', '=', 'IT'), ('code', 'like', 'machin')])
vj_tags = report_lines.expression_ids._get_matching_tags()
return bool(vj_tags & invoice_lines_tag_ids)

Something like that :p 

@Shawcker